Provides a working build infrastructure

bcornec opened the pull request:
- Make packages for the following distributions:
  debian 8/9, ubuntu 16.04/16.10/17.04, centos 7 and fedora 25
- Find a compromise around composer using COMPOSER_PHAR variable
- Copy the correct VERSION file in /etc/fossology for spec
- Fix TimeZone computation when links are used
- Adjusting file write attributes for rpm package installation
- Vagrant additions for docker based package building
- Remains to fix: Fedora 26 has a syntax change in dnf makecache (solved
  in project-builder.org 0.15.1)
- Fix PHP version usage in install procedure cf #920

Should allow to close #688 and #692 for now (while the composer aspect s far from being ideal)

maxhbr commented on pbconf/Vagrantfile:
```diff
@@ -61,9 +72,29 @@ EOF
+# cd /home/vagrant/prj/fossology
+mkdir -p /home/vagrant/composer/
+/home/vagrant/prj/fossology/utils/install_composer.sh /home/vagrant/composer/
+mkdir /home/vagrant/prj/fossology/src/vendor
```
Here it is (at least for me) necessary to have the `-p` parameter

maxhbr commented:
a simple `vagrant up` fails with the following log:
<pre>
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'centos/7'...
[...]
==> default: ++ <b>pb -p fossology -m centos-7-x86_64 prepve</b>
==> default: Project: fossology
==> default: Action: prepve
==> default: Packages: fossology
==> default: Packages: fossology
==> default: Use of uninitialized value in pattern match (m//) at /usr/share/perl5/vendor_perl/ProjectBuilder/VE.pm line 122.
==> default: Sources handled (VEprep): /home/vagrant/prj/fossology/pbdelivery/fossology-3.1.0.0.20171031105042.tar.gz /home/vagrant/prj/fossology/pbdelivery/fossology-3.1.0.0.20171031105042.pbconf.tar.gz /home/vagrant/prj/fossology/pbconf/fossology.pb /home/vagrant/prj/fossology/pbdelivery/3.1.0-0.20171031105042.g77415e80.pb /home/vagrant/.pbrc /home/vagrant/prj/fossology/pbdelivery/pbrc /home/vagrant/prj/fossology/pbdelivery/pbprep-centos-7-x86_64
==> default: Executing pbscript on ///home/pb/prj/fossology/pbdelivery if needed... 
==> default: OK
==> default: [centos-7-x86_64] [centos-7-x86_64] Sending build context to Docker daemon  14.61MB

==> default: [centos-7-x86_64] Step 1/16 : FROM pb:centos-7-x86_64-pb
==> default: [centos-7-x86_64]  ---> 5539dbbff224
[...]
==> default: [centos-7-x86_64] Step 15/16 : RUN cd ///home/pb/prj/fossology/pbdelivery ; perl ./pbprep-centos-7-x86_64 ; if [ '0' = '0' ]; then rm -f pbprep-centos-7-x86_64; fi
==> default: [centos-7-x86_64]  ---> Running in 986820e507e9
==> default: [centos-7-x86_64] ... Execution needed
==> default: [centos-7-x86_64] ==== Start of script for ve centos-7-x86_64 =====
==> default: [centos-7-x86_64] setting up date with /bin/true
==> default: [centos-7-x86_64] preping packages on ve...
==> default: [centos-7-x86_64] pb (aka project-builder.org) Version 0.14.6-2336
==> default: [centos-7-x86_64] 
<b>==> default: [centos-7-x86_64] Usage:
==> default: [centos-7-x86_64]     pb [-vhSq][-r pbroot][-p project][[-s script -a account -P port][-T
==> default: [centos-7-x86_64]     VEtype][-t [os-ver-arch]][-m os-ver-arch[,...]]][-g][-i image] <action>
==> default: [centos-7-x86_64]     [<pkg1> ...]
==> default: [centos-7-x86_64] 
==> default: [centos-7-x86_64]     pb [--verbose][--help][--man][--quiet][--snapshot][--revision
==> default: [centos-7-x86_64]     pbroot][--project project][[--script script --account account --port
==> default: [centos-7-x86_64]     port][-T VEtype][--target [os-ver-arch]][--machine
==> default: [centos-7-x86_64]     os-ver-arch[,...]]][--nographic][--image image][--rebuild] <action>
==> default: [centos-7-x86_64]     [<pkg1> ...]</b>
==> default: [centos-7-x86_64] 
==> default: [centos-7-x86_64] Project: fossology
==> default: [centos-7-x86_64] Action: build2prep
==> default: [centos-7-x86_64] INFO: ------ Starting to prepare build environement ------
==> default: [centos-7-x86_64] Downloading additional repository file http://ftp.project-builder.org/centos/7/x86_64/pb.repo... OK
==> default: [centos-7-x86_64] Adding yum repository... OK
==> default: [centos-7-x86_64] Downloading additional repository file https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm... [91mERROR running command (wget -O /tmp/pb.TXWKQGbRlB/epel-release-7-10.noarch.rpm https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm) with cwd=/home/pb/prj/fossology, pid=12 at /usr/share/perl5/ProjectBuilder/Base.pm line 178.
==> default: [centos-7-x86_64] 	ProjectBuilder::Base::pb_system('wget -O /tmp/pb.TXWKQGbRlB/epel-release-7-10.noarch.rpm https...', 'Downloading additional repository file https://dl.fedoraproje...') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 620
==> default: [centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo_gen('HASH(0x36fca30)', 'http://ftp.project-builder.org/centos/7/x86_64/pb.repo,https:...') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 593
==> default: [centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo_gen_conf('HASH(0x36fca30)', '/home/pb/prj/fossology/pbdelivery/pbrc', 'addrepo') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 548
==> default: [centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo('HASH(0x36fca30)') called at /usr/bin/pb line 1627
==> default: [centos-7-x86_64] 	main::pb_build2prep() called at /usr/bin/pb line 864
==> default: [centos-7-x86_64] [0mERROR: child (wget -O /tmp/pb.TXWKQGbRlB/epel-release-7-10.noarch.rpm https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm) cwd=/home/pb/prj/fossology exited with value 8
==> default: [centos-7-x86_64] --2017-10-31 10:51:54--  https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm
==> default: [centos-7-x86_64] Resolving dl.fedoraproject.org (dl.fedoraproject.org)... 209.132.181.23, 209.132.181.25, 209.132.181.24
==> default: [centos-7-x86_64] Connecting to dl.fedoraproject.org (dl.fedoraproject.org)|209.132.181.23|:443... connected.
==> default: [centos-7-x86_64] HTTP request sent, awaiting response... 404 Not Found
==> default: [centos-7-x86_64] 2017-10-31 10:51:55 ERROR 404: Not Found.
==> default: [centos-7-x86_64] 
==> default: [centos-7-x86_64]  ---> d387d5459559
==> default: [centos-7-x86_64] Removing intermediate container 986820e507e9
==> default: [centos-7-x86_64] Step 16/16 : RUN rm -rf ///home/pb/prj/fossology/pbdelivery/*
==> default: [centos-7-x86_64]  ---> Running in 4fb343bbae00
==> default: [centos-7-x86_64]  ---> 02ae6c1370eb
==> default: [centos-7-x86_64] Removing intermediate container 4fb343bbae00
==> default: [centos-7-x86_64] Successfully built 02ae6c1370eb
==> default: [centos-7-x86_64] Successfully tagged pb:centos-7-x86_64-pb-fossology
==> default: ++ pb -p fossology -m centos-7-x86_64 build2ve
==> default: Project: fossology
==> default: Action: build2ve
==> default: Packages: fossology
==> default: Packages: fossology
==> default: <b>Use of uninitialized value in pattern match (m//) at /usr/share/perl5/vendor_perl/ProjectBuilder/VE.pm line 122.</b>
==> default: Sources handled (VEbuild): /home/vagrant/prj/fossology/pbdelivery/fossology-3.1.0.0.20171031105042.tar.gz /home/vagrant/prj/fossology/pbdelivery/fossology-3.1.0.0.20171031105042.pbconf.tar.gz /home/vagrant/prj/fossology/pbconf/fossology.pb /home/vagrant/prj/fossology/pbdelivery/3.1.0-0.20171031105042.g77415e80.pb /home/vagrant/.pbrc /home/vagrant/prj/fossology/pbdelivery/pbrc /home/vagrant/prj/fossology/pbdelivery/pbbuild-centos-7-x86_64
==> default: Executing pbscript on ///home/pb/prj/fossology/pbdelivery if needed... 
<b>==> default: ERROR running command (/usr/bin/docker run --cidfile="/tmp/pb.yUJrSQByds/ctn-centos-7-x86_64.cid" --user pb --name centos-7-x86_64-VEbuild -v /home/vagrant/prj/fossology/pbbuild:/home/pb/prj/fossology/pbbuild-2 -v /tmp/pb.yUJrSQByds/vebuild:/home/pb/prj/fossology/pbdelivery-2 -w /home/pb/prj/fossology/pbdelivery-2 pb:centos-7-x86_64-pb-fossology /bin/bash /home/pb/prj/fossology/pbdelivery-2/pbbuild-centos-7-x86_64) with cwd=/home/vagrant/prj/fossology/pbbuild, pid=21463 at /usr/share/perl5/vendor_perl/ProjectBuilder/Base.pm line 178.
==> default: 	ProjectBuilder::Base::pb_system('/usr/bin/docker run --cidfile="/tmp/pb.yUJrSQByds/ctn-centos-...', 'Executing pbscript on ///home/pb/prj/fossology/pbdelivery if ...', 'verbose_[centos-7-x86_64] ') called at /usr/bin/pb line 2846
==> default: 	main::pb_send2target('VEbuild', '/home/vagrant/prj/fossology/pbdelivery/pbbuild-centos-7-x86_64', 'centos-7-x86_64', 0, 0, 0, 3) called at /usr/bin/pb line 3266
==> default: 	main::pb_launchv('/home/vagrant/prj/fossology/pbdelivery/pbbuild-centos-7-x86_64', 've', 'centos-7-x86_64', 'VEbuild', 3, 0, 0, 0, undef, ...) called at /usr/bin/pb line 3038
==> default: 	main::pb_parallel_launchv('HASH(0x1a41610)', 've', 'VEbuild', 3, undef, 0) called at /usr/bin/pb line 3482
==> default: 	main::pb_build2v('ve', 'build') called at /usr/bin/pb line 864
==> default: ERROR: child (/usr/bin/docker run --cidfile="/tmp/pb.yUJrSQByds/ctn-centos-7-x86_64.cid" --user pb --name centos-7-x86_64-VEbuild -v /home/vagrant/prj/fossology/pbbuild:/home/pb/prj/fossology/pbbuild-2 -v /tmp/pb.yUJrSQByds/vebuild:/home/pb/prj/fossology/pbdelivery-2 -w /home/pb/prj/fossology/pbdelivery-2 pb:centos-7-x86_64-pb-fossology /bin/bash /home/pb/prj/fossology/pbdelivery-2/pbbuild-centos-7-x86_64) cwd=/home/vagrant/prj/fossology/pbbuild exited with value 8</b>
==> default: [centos-7-x86_64] [centos-7-x86_64] total 14272
==> default: [centos-7-x86_64] drwxr-xr-x. 2 pb pb      226 Oct 31 10:51 .
==> default: [centos-7-x86_64] drwxr-xr-x. 1 pb pb       61 Oct 31 10:51 ..
==> default: [centos-7-x86_64] -rw-r--r--. 1 pb pb      951 Oct 31 10:51 .pbrc
==> default: [centos-7-x86_64] -rw-r--r--. 1 pb pb       51 Oct 31 10:51 3.1.0-0.20171031105042.g77415e80.pb
==> default: [centos-7-x86_64] -rw-r--r--. 1 pb pb    25961 Oct 31 10:51 fossology-3.1.0.0.20171031105042.pbconf.tar.gz
==> default: [centos-7-x86_64] -rw-r--r--. 1 pb pb 14564725 Oct 31 10:51 fossology-3.1.0.0.20171031105042.tar.gz
==> default: [centos-7-x86_64] -rw-r--r--. 1 pb pb     3289 Oct 31 10:51 fossology.pb
==> default: [centos-7-x86_64] -rw-r--r--. 1 pb pb     1580 Oct 31 10:51 pbbuild-centos-7-x86_64
==> default: [centos-7-x86_64] -rw-r--r--. 1 pb pb     3514 Oct 31 10:51 pbrc
==> default: [centos-7-x86_64] ... Execution needed
==> default: [centos-7-x86_64] ==== Start of script for ve centos-7-x86_64 =====
==> default: [centos-7-x86_64] setting up date with /bin/true
==> default: [centos-7-x86_64] building packages on ve...
==> default: [centos-7-x86_64] pb (aka project-builder.org) Version 0.14.6-2336
==> default: [centos-7-x86_64] 
<b>==> default: [centos-7-x86_64] Usage:
==> default: [centos-7-x86_64]     pb [-vhSq][-r pbroot][-p project][[-s script -a account -P port][-T
==> default: [centos-7-x86_64]     VEtype][-t [os-ver-arch]][-m os-ver-arch[,...]]][-g][-i image] <action>
==> default: [centos-7-x86_64]     [<pkg1> ...]
==> default: [centos-7-x86_64] 
==> default: [centos-7-x86_64]     pb [--verbose][--help][--man][--quiet][--snapshot][--revision
==> default: [centos-7-x86_64]     pbroot][--project project][[--script script --account account --port
==> default: [centos-7-x86_64]     port][-T VEtype][--target [os-ver-arch]][--machine
==> default: [centos-7-x86_64]     os-ver-arch[,...]]][--nographic][--image image][--rebuild] <action>
==> default: [centos-7-x86_64]     [<pkg1> ...]</b>
==> default: [centos-7-x86_64] 
==> default: [centos-7-x86_64] Project: fossology
==> default: [centos-7-x86_64] Action: build2pkg
==> default: [centos-7-x86_64] INFO: ------ Starting to prepare build environement ------
==> default: [centos-7-x86_64] Downloading additional repository file http://ftp.project-builder.org/centos/7/x86_64/pb.repo... OK
==> default: [centos-7-x86_64] Downloading additional repository file https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm... ERROR running command (wget -O /tmp/pb.smP7X3ROV2/epel-release-7-10.noarch.rpm https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm) with cwd=/home/pb/prj/fossology, pid=27 at /usr/share/perl5/ProjectBuilder/Base.pm line 178.
==> default: [centos-7-x86_64] 	ProjectBuilder::Base::pb_system('wget -O /tmp/pb.smP7X3ROV2/epel-release-7-10.noarch.rpm https...', 'Downloading additional repository file https://dl.fedoraproje...') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 620
==> default: [centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo_gen('HASH(0x32b6a60)', 'http://ftp.project-builder.org/centos/7/x86_64/pb.repo,https:...') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 593
==> default: [centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo_gen_conf('HASH(0x32b6a60)', '/home/pb/prj/fossology/pbdelivery/pbrc', 'addrepo') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 548
==> default: [centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo('HASH(0x32b6a60)') called at /usr/bin/pb line 1627
==> default: [centos-7-x86_64] 	main::pb_build2prep() called at /usr/bin/pb line 1685
==> default: [centos-7-x86_64] 	main::pb_build2pkg(undef) called at /usr/bin/pb line 864
<b>==> default: [centos-7-x86_64] ERROR: child (wget -O /tmp/pb.smP7X3ROV2/epel-release-7-10.noarch.rpm https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm) cwd=/home/pb/prj/fossology exited with value 8
==> default: [centos-7-x86_64] --2017-10-31 10:51:59--  https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm
==> default: [centos-7-x86_64] Resolving dl.fedoraproject.org (dl.fedoraproject.org)... 209.132.181.23, 209.132.181.25, 209.132.181.24
==> default: [centos-7-x86_64] Connecting to dl.fedoraproject.org (dl.fedoraproject.org)|209.132.181.23|:443... connected.
==> default: [centos-7-x86_64] HTTP request sent, awaiting response... 404 Not Found
==> default: [centos-7-x86_64] 2017-10-31 10:51:59 ERROR 404: Not Found.</b>
==> default: [centos-7-x86_64] 
<b>The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.</b>
</pre>



In the vagrant log there is often the following warning:
```
Use of uninitialized value in pattern match (m//) at /usr/share/perl5/vendor_perl/ProjectBuilder/VE.pm line 122. 
```

maxhbr commented on pbconf/fossology.pb:
```diff
@@ -46,14 +51,19 @@ sshport fossology = 22
+#
+#addrepo centos-5-x86_64 = http://packages.sw.be/rpmforge-release/rpmforge-release-0.3.6-1.el5.rf.x86_64.rpm,ftp://ftp.project-builder.org/test/centos/5/pb.repo
+#addrepo centos-4-x86_64 = http://packages.sw.be/rpmforge-release/rpmforge-release-0.3.6-1.el4.rf.x86_64.rpm,ftp://ftp.project-builder.org/test/centos/4/pb.repo
+addrepo centos-7-x86_64 = http://ftp.project-builder.org/centos/7/x86_64/pb.repo,https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm
```
The link `https://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7-10.noarch.rpm` returns 404

bcornec replied:
Seems Fedora updated again their rpm:
https://dl.fedoraproject.org/pub/epel/7/x86_64/Packages/e/epel-release-7-11.noarch.rpm

Normally the latest version should be used:
https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm

But I think we changed as we had an issue using this in the past (even if I don't remember why :-()

maxhbr replied:
If we want reproducible builds we have to pin versions.

bcornec replied:
Well, anyway we already use the generic version in the rest of the Vagrant setup for now. And that has no real impact on reproduceable builds (which we will never have anyway with composer downloaded on the fly, or packages installed from remote sources we don't control, so teir version may evolve). However not pinning makes maintenance much more easy.

bcornec commented:
I fixed the 2 remarks in an updated commit. Let me know if that works better for you @maxhbr

maxhbr commented:
@bcornec You have not yet updated the PR?

bcornec commented:
The rebase requested more manual fixes I envisonned sorry. Now pushed...

maxhbr commented:
@bcornec thank you

maxhbr commented:
The same step still fails with a different error: it complains that `
epel-release-7-11.noarch` is already installed:
<pre>
[...]
<b>++ pb -p fossology -m centos-7-x86_64 build2ve</b>
Project: fossology
Action: build2ve
Packages: fossology
Use of uninitialized value in pattern match (m//) at /usr/share/perl5/vendor_perl/ProjectBuilder/VE.pm line 122.
Packages: fossology
Sources handled (VEbuild): /home/vagrant/prj/fossology/pbdelivery/fossology-3.1.0.0.20171031162821.tar.gz /home/vagrant/prj/fossology/pbdelivery/fossology-3.1.0.0.20171031162821.pbconf.tar.gz /home/vagrant/prj/fossology/pbconf/fossology.pb /home/vagrant/prj/fossology/pbdelivery/3.1.0-0.20171031162821.ge05f0fab.pb /home/vagrant/.pbrc /home/vagrant/prj/fossology/pbdelivery/pbrc /home/vagrant/prj/fossology/pbdelivery/pbbuild-centos-7-x86_64
Executing pbscript on ///home/pb/prj/fossology/pbdelivery if needed... 
<b>ERROR running command (/usr/bin/docker run --cidfile="/tmp/pb.4equEx98Bd/ctn-centos-7-x86_64.cid" --user pb --name centos-7-x86_64-VEbuild -v /home/vagrant/prj/fossology/pbbuild:/home/pb/prj/fossology/pbbuild-2 -v /tmp/pb.4equEx98Bd/vebuild:/home/pb/prj/fossology/pbdelivery-2 -w /home/pb/prj/fossology/pbdelivery-2 pb:centos-7-x86_64-pb-fossology /bin/bash /home/pb/prj/fossology/pbdelivery-2/pbbuild-centos-7-x86_64) with cwd=/home/vagrant/prj/fossology/pbbuild, pid=21493 at /usr/share/perl5/vendor_perl/ProjectBuilder/Base.pm line 178.
	ProjectBuilder::Base::pb_system('/usr/bin/docker run --cidfile="/tmp/pb.4equEx98Bd/ctn-centos-...', 'Executing pbscript on ///home/pb/prj/fossology/pbdelivery if ...', 'verbose_[centos-7-x86_64] ') called at /usr/bin/pb line 2846
	main::pb_send2target('VEbuild', '/home/vagrant/prj/fossology/pbdelivery/pbbuild-centos-7-x86_64', 'centos-7-x86_64', 0, 0, 0, 3) called at /usr/bin/pb line 3266
	main::pb_launchv('/home/vagrant/prj/fossology/pbdelivery/pbbuild-centos-7-x86_64', 've', 'centos-7-x86_64', 'VEbuild', 3, 0, 0, 0, undef, ...) called at /usr/bin/pb line 3038
	main::pb_parallel_launchv('HASH(0x24ec610)', 've', 'VEbuild', 3, undef, 0) called at /usr/bin/pb line 3482
	main::pb_build2v('ve', 'build') called at /usr/bin/pb line 864
ERROR: child (/usr/bin/docker run --cidfile="/tmp/pb.4equEx98Bd/ctn-centos-7-x86_64.cid" --user pb --name centos-7-x86_64-VEbuild -v /home/vagrant/prj/fossology/pbbuild:/home/pb/prj/fossology/pbbuild-2 -v /tmp/pb.4equEx98Bd/vebuild:/home/pb/prj/fossology/pbdelivery-2 -w /home/pb/prj/fossology/pbdelivery-2 pb:centos-7-x86_64-pb-fossology /bin/bash /home/pb/prj/fossology/pbdelivery-2/pbbuild-centos-7-x86_64) cwd=/home/vagrant/prj/fossology/pbbuild exited with value 1</b>
[centos-7-x86_64] [centos-7-x86_64] total 14272
[centos-7-x86_64] drwxr-xr-x. 2 pb pb      226 Oct 31 16:30 .
[centos-7-x86_64] drwxr-xr-x. 1 pb pb       61 Oct 31 16:30 ..
[centos-7-x86_64] -rw-r--r--. 1 pb pb      951 Oct 31 16:30 .pbrc
[centos-7-x86_64] -rw-r--r--. 1 pb pb       51 Oct 31 16:30 3.1.0-0.20171031162821.ge05f0fab.pb
[centos-7-x86_64] -rw-r--r--. 1 pb pb    25940 Oct 31 16:30 fossology-3.1.0.0.20171031162821.pbconf.tar.gz
[centos-7-x86_64] -rw-r--r--. 1 pb pb 14564883 Oct 31 16:30 fossology-3.1.0.0.20171031162821.tar.gz
[centos-7-x86_64] -rw-r--r--. 1 pb pb     3282 Oct 31 16:30 fossology.pb
[centos-7-x86_64] -rw-r--r--. 1 pb pb     1580 Oct 31 16:30 pbbuild-centos-7-x86_64
[centos-7-x86_64] -rw-r--r--. 1 pb pb     3507 Oct 31 16:30 pbrc
[centos-7-x86_64] ... Execution needed
[centos-7-x86_64] ==== Start of script for ve centos-7-x86_64 =====
[centos-7-x86_64] setting up date with /bin/true
[centos-7-x86_64] building packages on ve...
[centos-7-x86_64] pb (aka project-builder.org) Version 0.14.6-2336
[centos-7-x86_64] 
[centos-7-x86_64] Usage:
[centos-7-x86_64]     pb [-vhSq][-r pbroot][-p project][[-s script -a account -P port][-T
[centos-7-x86_64]     VEtype][-t [os-ver-arch]][-m os-ver-arch[,...]]][-g][-i image] <action>
[centos-7-x86_64]     [<pkg1> ...]
[centos-7-x86_64] 
[centos-7-x86_64]     pb [--verbose][--help][--man][--quiet][--snapshot][--revision
[centos-7-x86_64]     pbroot][--project project][[--script script --account account --port
[centos-7-x86_64]     port][-T VEtype][--target [os-ver-arch]][--machine
[centos-7-x86_64]     os-ver-arch[,...]]][--nographic][--image image][--rebuild] <action>
[centos-7-x86_64]     [<pkg1> ...]
[centos-7-x86_64] 
[centos-7-x86_64] Project: fossology
[centos-7-x86_64] Action: build2pkg
[centos-7-x86_64] INFO: ------ Starting to prepare build environement ------
[centos-7-x86_64] Downloading additional repository file http://ftp.project-builder.org/centos/7/x86_64/pb.repo... OK
[centos-7-x86_64] Downloading additional repository file https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm... OK
[centos-7-x86_64] rpm -q --quiet epel-release-latest-7.noarch... NOT OK but non blocking
[centos-7-x86_64] Adding package epel-release-latest-7.noarch.rpm to setup repository... ERROR running command (sudo rpm -Uvh /tmp/pb.DO9KF1YYau/epel-release-latest-7.noarch.rpm) with cwd=/home/pb/prj/fossology, pid=27 at /usr/share/perl5/ProjectBuilder/Base.pm line 178.
[centos-7-x86_64] 	ProjectBuilder::Base::pb_system('sudo rpm -Uvh /tmp/pb.DO9KF1YYau/epel-release-latest-7.noarch...', 'Adding package epel-release-latest-7.noarch.rpm to setup repo...') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 631
[centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo_gen('HASH(0x33c5a20)', 'http://ftp.project-builder.org/centos/7/x86_64/pb.repo,https:...') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 593
[centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo_gen_conf('HASH(0x33c5a20)', '/home/pb/prj/fossology/pbdelivery/pbrc', 'addrepo') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 548
[centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo('HASH(0x33c5a20)') called at /usr/bin/pb line 1627
[centos-7-x86_64] 	main::pb_build2prep() called at /usr/bin/pb line 1685
[centos-7-x86_64] 	main::pb_build2pkg(undef) called at /usr/bin/pb line 864
[centos-7-x86_64] ERROR: <b>child (sudo rpm -Uvh /tmp/pb.DO9KF1YYau/epel-release-latest-7.noarch.rpm) cwd=/home/pb/prj/fossology exited with value 1</b>
[centos-7-x86_64] Preparing...                          ########################################
[centos-7-x86_64] 	<b>package epel-release-7-11.noarch is already installed</b>
</pre>

bcornec commented:
Ahhh ! you found why we moved to more specific rpm for EPEL :-(

If you install epel-release-latest-7.noarch.rpm as we did here, then rpm -q doesn't find that package (but could find epel-relrease-7.11.... instead) , so pb decides to install it, and it can't becasue it's already there :-(

IMHO this is a packaging error on EPEL side, but there is not much I can do for that at the moment, except going back to a precise version. That's why we changed it. Good to document here.

I'll see if checking with provides helps....

maxhbr commented on pbconf/Vagrantfile:
```diff
@@ -39,14 +40,21 @@ wget https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
 rpm -Uvh epel-release-latest-7.noarch.rpm
```
@bcornec maybe it would help if we would not install epel here?

maxhbr replied:
Oh, the problem is within a docker container.

maxhbr commented:
The following commet is based on the setting
```
DISTRO_LIST="ubuntu-14.10-x86_64 ubuntu-16.04-x86_64 debian-7-x86_64 debian-8-x86_64 centos-7-x86_64 fedora-25-x86_64" 
```

The result looks like:
```
/home/vagrant/repo
├── fedora
│   └── 25
│       └── x86_64
│           ├── fossology-3.1.0-0.20171101174246.ge05f0fab.fc25.src.rpm
│           ├── [...]
│           ├── fossology-wgetagent-3.1.0-0.20171101174246.ge05f0fab.fc25.x86_64.rpm
│           └── repodata
│               ├── [...]
│               └── repomd.xml
└── ubuntu
    ├── 16.04
    │   ├── fossology_3.1.0~rc2-1_amd64.changes
    │   ├──  [...]
    │   └── fossology-wgetagent_3.1.0~rc2-1_amd64.deb
    └── dists
        └── 16.04
            ├── Contents.gz
            ├── contrib
            │   ├── binary-all
            │   │   ├── Packages
            │   │   ├── Packages.bz2
            │   │   ├── Packages.gz
            │   │   └── Release
            │   ├── binary-amd64
            │   │   ├── Packages
            │   │   ├── Packages.bz2
            │   │   ├── Packages.gz
            │   │   └── Release
            │   └── source
            │       ├── Release
            │       ├── Sources
            │       ├── Sources.bz2
            │       └── Sources.gz
            └── Release

12 directories, 78 files
```
so building for `ubuntu-16.04` and `fedora-25` seem to be successful.


Building for debian 7 fails with the following error:
```
[..]
[debian-7-x86_64] ERROR running command (sudo /usr/bin/apt-get update ; sudo /usr/bin/env DEBIAN_FRONTEND=noninteractive /usr/bin/apt-get -y --allow-unauthenticated install  libglib2.0-dev libmagic-dev libxml2-dev libmxml-dev librpm-dev subversion rpm libpcre3-dev libssl-dev postgresql-server-dev-all libboost-regex-dev libboost-program-options-dev php-mbstring php-xml PBPHPCLI curl postgresql-server-dev-8.4) with cwd=/home/pb/prj/fossology, pid=25 at /usr/share/perl/5.14/ProjectBuilder/Base.pm line 178
[debian-7-x86_64]       ProjectBuilder::Base::pb_system('sudo /usr/bin/apt-get update ; sudo /usr/bin/env DEBIAN_FRONT...', 'Re-trying installing dependencies (sudo /usr/bin/apt-get upda...') called at /usr/share/perl/5.14/ProjectBuilder/Distribution.pm line 365
[debian-7-x86_64]       ProjectBuilder::Distribution::pb_distro_installdeps('/tmp/pb.84DRpn5v__/control', 'HASH(0x3889600)', undef) called at /usr/bin/pb line 1670
[debian-7-x86_64]       main::pb_build2prep() called at /usr/bin/pb line 1685
[debian-7-x86_64]       main::pb_build2pkg(undef) called at /usr/bin/pb line 864
[debian-7-x86_64] ERROR: child (sudo /usr/bin/apt-get update ; sudo /usr/bin/env DEBIAN_FRONTEND=noninteractive /usr/bin/apt-get -y --allow-unauthenticated install  libglib2.0-dev libmagic-dev libxml2-dev libmxml-dev librpm-dev subversion rpm libpcre3-dev libssl-dev postgresql-server-dev-all libboost-regex-dev libboost-program-options-dev php-mbstring php-xml PBPHPCLI curl postgresql-server-dev-8.4) cwd=/home/pb/prj/fossology exited with value 100
[debian-7-x86_64] Reading package lists...
[debian-7-x86_64] Building dependency tree...
[debian-7-x86_64] Reading state information...
[debian-7-x86_64] E: Unable to locate package php-mbstring
[debian-7-x86_64] E: Unable to locate package php-xml
[debian-7-x86_64] E: Unable to locate package PBPHPCLI
[debian-7-x86_64] E: Unable to locate package postgresql-server-dev-8.4
[debian-7-x86_64] E: Couldn't find any package by regex 'postgresql-server-dev-8.4'
```

Building for debian 8 fails with the following error:
```
[...]
[debian-8-x86_64] Reading package lists...
[debian-8-x86_64] ERROR: child (sudo /usr/bin/apt-get update ; sudo /usr/bin/env DEBIAN_FRONTEND=noninteractive /usr/bin/apt-get -y --allow-unauthenticated install  libglib2.0-dev libmagic-dev libxml2-dev libmxml-dev librpm-dev subversion rpm libpcre3-dev libssl-dev postgresql-server-dev-all libboost-regex-dev libboost-program-options-dev php-mbstring php-xml php5-cli curl postgresql-server-dev-9.4 libjson0-dev) cwd=/home/pb/prj/fossology exited with value 100
[debian-8-x86_64] Reading package lists...
[debian-8-x86_64] Building dependency tree...
[debian-8-x86_64] Reading state information...
[debian-8-x86_64] E: Unable to locate package php-mbstring
[debian-8-x86_64] E: Unable to locate package php-xml
[debian-8-x86_64] ERROR running command (sudo /usr/bin/apt-get update ; sudo /usr/bin/env DEBIAN_FRONTEND=noninteractive /usr/bin/apt-get -y --allow-unauthenticated install  libglib2.0-dev libmagic-dev libxml2-dev libmxml-dev librpm-dev subversion rpm libpcre3-dev libssl-dev postgresql-server-dev-all libboost-regex-dev libboost-program-options-dev php-mbstring php-xml php5-cli curl postgresql-server-dev-9.4 libjson0-dev) with cwd=/home/pb/prj/fossology, pid=25 at /usr/share/perl/5.20/ProjectBuilder/Base.pm line 178.
[debian-8-x86_64] 	ProjectBuilder::Base::pb_system("sudo /usr/bin/apt-get update ; sudo /usr/bin/env DEBIAN_FRONT"..., "Re-trying installing dependencies (sudo /usr/bin/apt-get upda"...) called at /usr/share/perl/5.20/ProjectBuilder/Distribution.pm line 365
[debian-8-x86_64] 	ProjectBuilder::Distribution::pb_distro_installdeps("/tmp/pb.Sb7jEp1ETs/control", HASH(0x38a9d88), undef) called at /usr/bin/pb line 1670
[debian-8-x86_64] 	main::pb_build2prep() called at /usr/bin/pb line 1685
[debian-8-x86_64] 	main::pb_build2pkg(undef) called at /usr/bin/pb line 864
```

Building for centos 7 fails with the following error (already mentioned above:
```
[centos-7-x86_64] rpm -q --quiet epel-release-latest-7.noarch... NOT OK but non blocking
[centos-7-x86_64] Adding package epel-release-latest-7.noarch.rpm to setup repository... ERROR: child (sudo rpm -Uvh /tmp/pb.oxrHPIc4vJ/epel-release-latest-7.noarch.rpm) cwd=/home/pb/prj/fossology exited with value 1
[centos-7-x86_64] Preparing...                          ########################################
[centos-7-x86_64] 	package epel-release-7-11.noarch is already installed
[centos-7-x86_64] ERROR running command (sudo rpm -Uvh /tmp/pb.oxrHPIc4vJ/epel-release-latest-7.noarch.rpm) with cwd=/home/pb/prj/fossology, pid=27 at /usr/share/perl5/ProjectBuilder/Base.pm line 178.
[centos-7-x86_64] 	ProjectBuilder::Base::pb_system('sudo rpm -Uvh /tmp/pb.oxrHPIc4vJ/epel-release-latest-7.noarch...', 'Adding package epel-release-latest-7.noarch.rpm to setup repo...') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 631
[centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo_gen('HASH(0x26b2cb0)', 'http://ftp.project-builder.org/centos/7/x86_64/pb.repo,https:...') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 593
[centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo_gen_conf('HASH(0x26b2cb0)', '/home/pb/prj/fossology/pbdelivery/pbrc', 'addrepo') called at /usr/share/perl5/ProjectBuilder/Distribution.pm line 548
[centos-7-x86_64] 	ProjectBuilder::Distribution::pb_distro_setuprepo('HASH(0x26b2cb0)') called at /usr/bin/pb line 1627
[centos-7-x86_64] 	main::pb_build2prep() called at /usr/bin/pb line 1685
[centos-7-x86_64] 	main::pb_build2pkg(undef) called at /usr/bin/pb line 864
```

maxhbr commented:
@bcornec:
- can you give me a list of supported platforms, e.g. which packages should build sucessfully?
- is the simple vagrant environment good enough for testing?
  - can you recommend a better environment for testing?

maxhbr commented on pbconf/Vagrantfile:
```diff
@@ -23,7 +23,8 @@ Vagrant.configure("2") do |config|
 
   config.vm.synced_folder "..", "/home/vagrant/prj/fossology", type: "rsync"
-  config.vm.network "forwarded_port", guest: 80, host: 8081
+  config.vm.synced_folder "../target", "/home/vagrant/repo", type: "rsync", create: true
```
It might be better if that is not of type rsync. At least for me the rsync does not automatically work.

maxhbr commented:
I have tried to install the generated ubuntu 16.04 packages with the following vagrantfile but I ran into errors.
```
Vagrant.configure("2") do |config|
  config.vm.box = "ubuntu/xenial64"
  config.vm.synced_folder "../../target", "/home/vagrant/target", create: true
  config.vm.provision "shell" do |s|
    s.privileged = false
    s.inline = <<SCRIPT
set -ex
cd /home/vagrant/target/ubuntu/16.04/
sudo dpkg -i *.deb || sudo apt-get install -y -f
SCRIPT
  end
end
```

The log output is:
<pre>
[...]
++ sudo apt-get install -y -f
[...]
update-alternatives: error: 
no alternatives for upx
update-alternatives: 
using /usr/bin/upx-ucl to provide /usr/bin/upx (upx) in auto mode
Setting up unrar-free (1:0.0.1+cvs20140707-1) ...
update-alternatives: 
using /usr/bin/unrar-free to provide /usr/bin/unrar (unrar) in auto mode
Setting up unzip (6.0-20ubuntu1) ...
Setting up p7zip-full (9.20.1~dfsg.1-4.2) ...
Setting up p7zip (9.20.1~dfsg.1-4.2) ...
Setting up libboost-program-options1.58.0:amd64 (1.58.0+dfsg-5ubuntu3.1) ...
Setting up libboost-regex1.58.0:amd64 (1.58.0+dfsg-5ubuntu3.1) ...
Setting up fossology-copyright (3.1.0~rc2-1) ...
Setting up fossology-nomos (3.1.0~rc2-1) ...
Setting up fossology-mimetype (3.1.0~rc2-1) ...
Setting up fossology-delagent (3.1.0~rc2-1) ...
Setting up libserf-1-1:amd64 (1.3.8-1) ...
Setting up libsvn1:amd64 (1.9.3-2ubuntu1.1) ...
Setting up subversion (1.9.3-2ubuntu1.1) ...
Setting up fossology-wgetagent (3.1.0~rc2-1) ...
<b>dpkg: dependency problems prevent configuration of fossology:
 fossology depends on fossology-web; however:
  Package fossology-web is not configured yet.
dpkg: error processing package fossology (--configure):
 dependency problems - leaving unconfigured</b>
Setting up fossology-debug (3.1.0~rc2-1) ...
No apport report written because the error message indicates its a followup error from a previous failure.
Setting up fossology-deciderjob (3.1.0~rc2-1) ...
Setting up fossology-dev (3.1.0~rc2-1) ...
Setting up fossology-monk (3.1.0~rc2-1) ...
Setting up fossology-monkbulk (3.1.0~rc2-1) ...
Setting up fossology-readmeoss (3.1.0~rc2-1) ...
Setting up fossology-reuser (3.1.0~rc2-1) ...
Setting up fossology-spdx2 (3.1.0~rc2-1) ...
Setting up fossology-unifiedreport (3.1.0~rc2-1) ...
Setting up gnustep-common (2.6.6-3) ...
Setting up gnustep-base-common (1.24.7-1build2) ...
Setting up libavahi-common-data:amd64 (0.6.32~rc+dfsg-1ubuntu2) ...
Setting up libavahi-common3:amd64 (0.6.32~rc+dfsg-1ubuntu2) ...
Setting up libavahi-client3:amd64 (0.6.32~rc+dfsg-1ubuntu2) ...
Setting up libobjc4:amd64 (5.4.0-6ubuntu1~16.04.5) ...
Setting up libgnustep-base1.24 (1.24.7-1build2) ...
Setting up gnustep-base-runtime (1.24.7-1build2) ...
Setting up libwavpack1:amd64 (4.75.2-2) ...
Setting up unar (1.9.1-1) ...
Setting up libsensors4:amd64 (1:3.4.0-2) ...
Setting up postgresql-contrib-9.5 (9.5.9-0ubuntu0.16.04) ...
Setting up sysstat (11.2.0-1ubuntu0.1) ...

Creating config file /etc/default/sysstat with new version
update-alternatives: using /usr/bin/sar.sysstat to provide /usr/bin/sar (sar) in auto mode
Setting up libnss3-nssdb (2:3.28.4-0ubuntu0.16.04.3) ...
Setting up libnss3:amd64 (2:3.28.4-0ubuntu0.16.04.3) ...
Setting up librpmio3 (4.12.0.1+dfsg1-3build3) ...
Setting up librpm3 (4.12.0.1+dfsg1-3build3) ...
Setting up librpmbuild3 (4.12.0.1+dfsg1-3build3) ...
Setting up librpmsign3 (4.12.0.1+dfsg1-3build3) ...
Setting up rpm-common (4.12.0.1+dfsg1-3build3) ...
Setting up rpm2cpio (4.12.0.1+dfsg1-3build3) ...
Setting up debugedit (4.12.0.1+dfsg1-3build3) ...
Setting up rpm (4.12.0.1+dfsg1-3build3) ...
Setting up fossology-ununpack (3.1.0~rc2-1) ...
Setting up fossology-pkgagent (3.1.0~rc2-1) ...
Setting up fossology-buckets (3.1.0~rc2-1) ...
Processing triggers for libc-bin (2.23-0ubuntu9) ...
Processing triggers for systemd (229-4ubuntu19) ...
Processing triggers for ureadahead (0.100.0-19) ...
Processing triggers for ufw (0.35-0ubuntu2) ...
<b>Errors were encountered while processing:
 fossology-web
 fossology
E: Sub-process /usr/bin/dpkg returned an error code (1)</b>
</pre>

bcornec commented:
Reason of the problem for CentOS7:
```
 $ docker run -ti pb:centos-7-x86_64
[root@28cadcb5499b /]# rpm -aq | grep epel
[root@28cadcb5499b /]# rpm -Uvh https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
Retrieving https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
warning: /var/tmp/rpm-tmp.XhDJDA: Header V3 RSA/SHA256 Signature, key ID 352c64e5: NOKEY
Preparing...                          ################################# [100%]
Updating / installing...
   1:epel-release-7-11                ################################# [100%]
[root@28cadcb5499b /]# rpm -aq | grep epel
epel-release-7-11.noarch
```
So the package doesn't provide its own name in the RPM DB. Thus when pb tests for pre-existence of the package epel-release-latest, it doesn't find it, tries to install again, and has an error because it's already installed.

I reported it upstream in EPEL as https://bugzilla.redhat.com/show_bug.cgi?id=1509190
Meanwhile we can use the precise version. Will update patch for this.

I have removed the non-supported distros (such as Debian 7 and Ubuntu 14.04 from the list - still commented for now).

For Debian 8, php-cli already contains the mbstring and xml extensions so I modified the build files accordingly (I also made changes for Ubuntu which will need checks)

For me the Vagrant env is fine for testing.

With regards to the installation, what should be done is download the file generated during the pb run for the correct distro, and add it to your test environment in the /etc/apt/sources.list directory, then run apt-get upgrade && apt-get install fossology. That's how it's intended to work (also for your users BTW).

mcjaeger commented:
I think it is working now and we should move more issues to another PR as this is improving a lot and we have always the effort to rebase with master.

bcornec commented:
+1 !

mcjaeger commented:
(remark: I think the basic issue is not to install fossology-ninka)

maxhbr commented:
Do all the tests with vagrant mentioned above succeed?
